### ISSUE TITLE
System.OperatingSystem should include Proton

### ISSUE THREAD

**Issue by Crystal-Shard** [adventuregamestudio/ags#2167]
**Describe the problem**
Steam supports the "Steam Deck", is a handheld computer that plays numerous games. However, it has different input devices than a standard computer, i.e. a touchscreen and controller joypad/buttons.
So it would be great if AGS could detect, at run time, whether it is running on a computer or on a portable device, so the programmer can (e.g.) enable or disable certain GUI elements accordingly. The Steam Deck runs on Linux and supports windows games via Proton (which is a fork of WINE), using the same game executables as in windows.

**Suggested change**
When running in Proton, System.OperatingSystem should return eOSProton. Googling the issue suggests one of these code snippets would straightforwardly detect that.
- https://www.construct.net/en/forum/construct-3/how-do-i-8/detect-game-running-steam-deck-177909
- https://www.winehq.org/pipermail/wine-devel/2008-September/069387.html

Perhaps alternatively, detecting whether a keyboard or mouse is present could work.

**Comment by AlanDrake** [comment]
Proton can run on any kind of linux machine, so it's not enough to detect a Steam Deck.
A Steam Deck could also run the Linux version natively.

It might be possible to fetch the system vendor/model values, but the official way is using steam works API, so perhaps a plugin is the correct way in this case?

**Comment by ivan-mogilko** [comment]
The situation with OS detection in AGS now is that it does not really detect anything at runtime, but relies on engine's compile-time flags. In simpler words this means that this info is "embedded" when you compile engine for particular platform.

In order to change this we'd have to implement runtime detection probably, at least to detect some extra details about current system. Possibly we would also have to add additional fields to System, so to not fill OperatingSystem with all possible sub-types, as that might make this property inconvenient to work with.

I am speaking hypothetically now, as I never looked into this.

I think that it may make sense to add device detection as a side-feature, as that might also be useful in any case.

I suppose that a workaround could be is to add some custom file, or a config option that you place into acsetup.cfg yourself before shipping the game for Steam Deck and read back in script, **IF** you ship for Steam Deck exclusively (if you don't, then it won't work of course).

**Comment by ivan-mogilko** [comment]
Something I might add, in relation to this, and recently came up issue with vsync, where I mentioned that AGS does not provide an option to know whether vsync is supported.

In terms of script API, I suppose, it may be more convenient to add not individual properties for system info, but something that may be extended easily, such as: a function like GetProperty (GetCaps etc), and indexed attribute (Properties[], Caps[]), or similar.
("Caps" stands for "Capabilities", this term is used in some device SDKs).

**Comment by Crystal-Shard** [comment]
Right. The issue here is that Steam allows you to have different builds for Windows / Linux / Mac, but does not allow a different build for the SteamDeck; in the case of AGS games, it will use the Windows build under Proton. This is why I'm looking for a way to check at runtime.

**Comment by ivan-mogilko** [comment]
To summarize, we'd need to:

* Investigate how to detect necessary OS properties at runtime;
* Design a convenient script API for telling these properties, as pushing everything into the OperatingSystem enum seems not to be a very good idea.

Alternatively:

* Find out if it's possible to detect keyboard and mouse presence, perhaps using SDL2, or, if that is not supported by the library, then other means per-system.
* Design a simple script API for telling this, like `System.HasMouse`, `System.HasKeyboard` or similar.
* If we go this route, perhaps detecting presence of touch screen and gamepad controls may also be useful. As mentioned, we have joystick support in ags4, but we could at least do a device presence detection if that's absolutely necessary.

Overall it seems like detecting available input devices may be a more suitable approach, rather than detecting particular system, as with OS properties you might need to keep the checks updated with every new system you support, while with controls detection you might do this once, and that will keep working for other enviroments with similar lack of certain controls (or presence of particular controls).

**Comment by Crystal-Shard** [comment]
@morganwillcock That Steamworks API function is an excellent solution.

**Comment by ericoporto** [comment]
This looks like something to be implemented in a plugin. 

The little OS check we had caused things like "Gemini Rue doesn't have rain in Linux", so anything that is not a feature support detection may be misused in the ags3 branch - if something for platform detection is done instead it's best on ags4.

**Comment by Crystal-Shard** [comment]
Yes, it would fit in the AGSteam.dll plugin, but as far as I'm aware it hasn't been in development for a couple years now.
